File: tutor-claude/jpa-tutor/src/main/java/com/jpatutor/ch15_performance/OrderQueryService.java
package com.jpatutor.ch15_performance;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * [15장] OSIV(Open Session In View)를 끈 상태(이 프로젝트의 application.yml: open-in-view: false)에서
 * "트랜잭션이 끝나면 영속성 컨텍스트(세션)도 함께 닫힌다"는 것을 보여주기 위한 서비스.
 *
 * OSIV가 true(스프링 부트 기본값)이면 컨트롤러가 응답을 렌더링할 때까지 영속성 컨텍스트가
 * 열려있어서, 트랜잭션이 끝난 뒤(서비스 메서드 반환 후)에도 지연 로딩을 계속 사용할 수 있다.
 * 하지만 이건 "요청-응답 전체" 동안 DB 커넥션을 하나 붙잡고 있다는 뜻이라, 트래픽이 많은
 * 서비스에서는 커넥션 풀 고갈로 이어질 수 있다. 그래서 실무에서는 open-in-view: false로 끄고,
 * 대신 "트랜잭션 안에서 필요한 데이터를 다 준비해서" DTO나 fetch join으로 반환하는 패턴을 쓴다.
 */
@Service
@RequiredArgsConstructor
public class OrderQueryService {

    private final OrderRepository orderRepository;

    // 지연 로딩 필드(member, orderItems)를 트랜잭션 안에서 전혀 건드리지 않고 그대로 반환한다.
    // -> 트랜잭션이 끝나는 순간 세션이 닫히므로, 반환된 프록시들은 더 이상 초기화할 수 없다.
    @Transactional(readOnly = true)
    public Order findOrderWithoutTouchingLazyFields(Long id) {
        return orderRepository.findById(id).orElseThrow();
    }

    // fetch join으로 트랜잭션 "안에서" 필요한 연관관계를 전부 미리 초기화해둔다.
    // -> 세션이 닫혀도 이미 채워진 값이라 트랜잭션 밖에서 접근해도 예외가 나지 않는다.
    @Transactional(readOnly = true)
    public Order findOrderWithFetchJoin(Long id) {
        return orderRepository.findByIdFetchAll(id).orElseThrow();
    }
}

File: tutor-claude/springboot-tutor/src/main/kotlin/be/com/springbootclaude/ddd/order/application/OrderQueryService.kt
package be.com.springbootclaude.ddd.order.application

import be.com.springbootclaude.ddd.order.domain.model.Money
import be.com.springbootclaude.ddd.order.domain.model.Order
import be.com.springbootclaude.ddd.order.domain.model.OrderStatus
import be.com.springbootclaude.ddd.order.domain.repository.OrderRepository
import be.com.springbootclaude.ddd.order.domain.service.OrderPriceCalculator
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

/**
 * Application Service: OrderQueryService
 *
 * CQRS의 Query Side:
 * - 읽기 전용 (조회만)
 * - 도메인 모델을 DTO로 변환
 * - 성능 최적화에 유리 (필요한 데이터만 조회)
 *
 * @Transactional(readOnly = true):
 * - 읽기 전용 트랜잭션
 * - 성능 최적화 (Dirty Checking 안 함)
 */
@Service
@Transactional(readOnly = true)
class OrderQueryService(
    private val orderRepository: OrderRepository,
    private val orderPriceCalculator: OrderPriceCalculator
) {

    /**
     * 조회: 주문 상세
     */
    fun getOrderDetail(orderId: Long): OrderDetailDto {
        val order = orderRepository.findById(orderId)
            ?: throw IllegalArgumentException("주문을 찾을 수 없습니다: $orderId")

        return OrderDetailDto.from(order, orderPriceCalculator)
    }

    /**
     * 조회: 고객의 주문 목록
     */
    fun getCustomerOrders(customerId: Long): List<OrderSummaryDto> {
        val orders = orderRepository.findByCustomerId(customerId)
        return orders.map { OrderSummaryDto.from(it) }
    }

    /**
     * 조회: 특정 상태의 주문 목록
     */
    fun getOrdersByStatus(status: OrderStatus): List<OrderSummaryDto> {
        val orders = orderRepository.findByStatus(status)
        return orders.map { OrderSummaryDto.from(it) }
    }

    /**
     * 조회: 기간별 주문 목록
     */
    fun getOrdersByPeriod(
        startDate: LocalDateTime,
        endDate: LocalDateTime
    ): List<OrderSummaryDto> {
        val orders = orderRepository.findByOrderedAtBetween(startDate, endDate)
        return orders.map { OrderSummaryDto.from(it) }
    }
}

/**
 * DTO: 주문 상세 정보
 * - Presentation Layer에 반환할 데이터
 * - 도메인 모델과 API 응답 분리
 */
data class OrderDetailDto(
    val orderId: Long,
    val customerId: Long,
    val status: OrderStatus,
    val deliveryAddress: String,
    val items: List<OrderItemDto>,
    val itemsTotal: Money,
    val deliveryFee: Money,
    val discount: Money,
    val finalAmount: Money,
    val earnedPoints: Int,
    val orderedAt: LocalDateTime
) {
    companion object {
        fun from(order: Order, priceCalculator: OrderPriceCalculator): OrderDetailDto {
            return OrderDetailDto(
                orderId = order.id!!,
                customerId = order.customerId,
                status = order.status,
                deliveryAddress = order.deliveryAddress.getFullAddress(),
                items = order.orderItems.map { OrderItemDto.from(it) },
                itemsTotal = order.calculateTotalAmount(),
                deliveryFee = priceCalculator.calculateDeliveryFee(order),
                discount = Money.ZERO, // 간단히 하기 위해 0으로
                finalAmount = priceCalculator.calculateFinalAmount(order),
                earnedPoints = priceCalculator.calculateEarnedPoints(order),
                orderedAt = order.orderedAt
            )
        }
    }
}

data class OrderItemDto(
    val productId: Long,
    val productName: String,
    val price: Money,
    val quantity: Int,
    val totalPrice: Money
) {
    companion object {
        fun from(orderItem: be.com.springbootclaude.ddd.order.domain.model.OrderItem): OrderItemDto {
            return OrderItemDto(
                productId = orderItem.productInfo.productId,
                productName = orderItem.productInfo.productName,
                price = orderItem.productInfo.price,
                quantity = orderItem.quantity,
                totalPrice = orderItem.calculateTotalPrice()
            )
        }
    }
}

/**
 * DTO: 주문 요약 정보
 * - 목록 조회용 (상세 정보보다 가벼움)
 */
data class OrderSummaryDto(
    val orderId: Long,
    val customerId: Long,
    val status: OrderStatus,
    val itemCount: Int,
    val totalAmount: Money,
    val orderedAt: LocalDateTime
) {
    companion object {
        fun from(order: Order): OrderSummaryDto {
            return OrderSummaryDto(
                orderId = order.id!!,
                customerId = order.customerId,
                status = order.status,
                itemCount = order.orderItems.size,
                totalAmount = order.calculateTotalAmount(),
                orderedAt = order.orderedAt
            )
        }
    }
}
